sys/linux/init_images.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package linux
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/google/syzkaller/prog"
+)
+
+func (arch *arch) extractSyzMountImage(c *prog.Call) (io.Reader, error) {
+	// In order to reduce the size of syzlang programs, disk imags are de facto compressed.
+	// Here we do the uncompression.
+	if c.Meta.CallName != "syz_mount_image" {
+		return nil, nil
+	}
+	ret, err := parseSyzMountImage(c)
+	if err != nil {
+		// Parsing failed --> do not try to recover, just ignore.
+		return nil, err
+	}
+	readers := []io.Reader{}
+	nextPos := 0
+	// Add fake zero readers between segments, so that we can combine them to read the whole image.
+	for _, segment := range ret.segments {
+		offset := int(segment.offset.Val)
+		if offset > nextPos {
+			readers = append(readers, &zeroReader{left: offset - nextPos})
+		}
+		size := int(segment.size.Val)
+		readers = append(readers, bytes.NewReader(segment.data.Data()[0:size]))
+		nextPos = offset + size
+	}
+	if int(ret.size.Val) > nextPos {
+		readers = append(readers, &zeroReader{left: int(ret.size.Val) - nextPos})
+	}
+	return io.MultiReader(readers...), nil
+}
+
+type zeroReader struct {
+	left int
+}
+
+func (zr *zeroReader) Read(p []byte) (n int, err error) {
+	if zr.left == 0 {
+		return 0, io.EOF
+	}
+	toRead := zr.left
+	if toRead > len(p) {
+		toRead = len(p)
+	}
+	for i := 0; i < toRead; i++ {
+		p[i] = 0
+	}
+	zr.left -= toRead
+	return toRead, nil
+}
+
+const imageMaxSize = 129 << 20
+
+func fixUpImageSegments(parsed *mountImageArgs) {
+	parsed.filterSegments(func(i int, segment *mountImageSegment) bool {
+		if segment.parseError != nil {
+			return false
+		}
+		return segment.offset.Val < imageMaxSize && segment.size.Val < imageMaxSize
+	})
+	// Overwriting of the image multiple times is not efficient and complicates image extraction in Go.
+	// So let's make segments non-overlapping.
+	sort.Stable(&sortSegments{parsed: parsed})
+	resizeSegment := func(s *mountImageSegment, size uint64) {
+		s.size.Val = size
+		s.data.SetData(s.data.Data()[0:size])
+	}
+
+	newSize := parsed.size.Val
+	for idx, segment := range parsed.segments {
+		actualSize := uint64(len(segment.data.Data()))
+		if segment.size.Val != actualSize {
+			segment.size.Val = actualSize
+		}
+		if idx > 0 {
+			// Adjust the end of the previous segment.
+			prevSegment := parsed.segments[idx-1]
+			if prevSegment.offset.Val+prevSegment.size.Val > segment.offset.Val {
+				resizeSegment(prevSegment, segment.offset.Val-prevSegment.offset.Val)
+			}
+		}
+		if segment.offset.Val+segment.size.Val > imageMaxSize {
+			resizeSegment(segment, imageMaxSize-segment.offset.Val)
+		}
+		if segment.offset.Val+segment.size.Val > newSize {
+			newSize = segment.offset.Val + segment.size.Val
+		}
+	}
+	if newSize > imageMaxSize {
```

Comment:
Is this possible? It shouldn't, right? Otherwise we still have segments that don't fit into imageMaxSize, not sure what will happen during serialization of such images.

Reply:
Indeed. Replaced with an assert.